modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java

```diff
@@ -1549,70 +1547,201 @@ protected void insert(SimpleFeature feature, SimpleFeatureType featureType, Conn
-                
-                for (Iterator f = features.iterator(); f.hasNext();) {
-                    SimpleFeature feature = (SimpleFeature) f.next();
-                    
-                    List<Object> keyValues = null;
-                    boolean useExisting = Boolean.TRUE.equals(feature.getUserData().get(Hints.USE_PROVIDED_FID));
-                    if(useExisting) {
-                        keyValues = decodeFID(key, feature.getID(), true);
-                    } else if (!postInsert) {
-                        keyValues = getNextValues( key, cx );
+            } catch (SQLException e) {
+                String msg = "Error inserting features";
+                throw (IOException) new IOException(msg).initCause(e);
+            }
+        }
+    }
+
+    /**
+     * Specialized insertion for dialects that are using prepared statements.
+     */
+    private void insertPS(Collection<SimpleFeature> features, InsertionClassifier kind,
+                          SimpleFeatureType featureType, Connection cx, PrimaryKey key)
+            throws IOException, SQLException {
+        final PreparedStatementSQLDialect dialect = (PreparedStatementSQLDialect) getSQLDialect();
+
+        final KeysFetcher keysFetcher = KeysFetcher.create(this, cx,kind.useExisting, key);
+
+        final String sql = buildInsertPS(kind, featureType, keysFetcher, dialect);
+        LOGGER.log(Level.FINE, "Inserting new features with ps: {0}", sql);
+
+        //create the prepared statement
+        final PreparedStatement ps;
+        if (keysFetcher.isPostInsert()) {
+            // ask the DB to return the values of all the keys after the insertion
+            ps = cx.prepareStatement(sql, keysFetcher.getColumnNames());
+        } else {
+            ps = cx.prepareStatement(sql);
+        }
+        try {
+            for (SimpleFeature feature : features) {
+                //set the attribute values
+                int i = 1;
+                for(AttributeDescriptor att : featureType.getAttributeDescriptors()) {
+                    String colName = att.getLocalName();
+                    // skip the pk columns in case we have exposed them, we grab the
+                    // value from the pk itself
+                    if(keysFetcher.isKey(colName)) {
+                        continue;
                     }
-                    
 
-                    if ( dialect instanceof PreparedStatementSQLDialect ) {
-                        PreparedStatement ps = insertSQLPS( featureType, feature, keyValues, cx );
-                        try {
-                            ((PreparedStatementSQLDialect)dialect).onInsert(ps, cx, featureType);
-                            ps.execute();
-                        } finally {
-                            closeSafe( ps );
+                    Class binding = att.getType().getBinding();
+
+                    Object value = feature.getAttribute(colName);
+                    /*if (value == null) {
```

Comment:
I guess it depends on the reviewer.. I honestly don't care to see the steps you followed to get to a result, it's a waste of time to me, I always review the final product, not the single commits.